### ISSUE TITLE
Default Game Keys

### ISSUE THREAD

**Issue by rkitover** [visualboyadvance-m/visualboyadvance-m#362]
I want to get some feedback from users about your preferred game keys (and any other keys you use) to see if changing the defaults makes sense.

Here is what I propose:

| keys | GBA keys |
|------|-----------|
| WASD | direction
| `[]` | a / b |
| P\ | l / r |
| enter | start |
| backspace | select |

other default keys like speedup would be somewhere in the `[]` area as well.

Also want to change the default joystick binds to match the 360/xbone controller.

Please let me know what you think.

**Comment by Rambin** [comment]
It depends on what game I'm playing.
For example, I use WSAD for direction and [K][J][L][;] for ABLR when playing Metroid, and [J][K][U][I] for ABLR when playing Gyakuten Saiban.
Is it possible to save key setting seperately for a game?

**Comment by rkitover** [comment]
That sounds like a nice feature to add, or perhaps configuration profiles of some sort.

**Comment by RunisDauphin** [comment]
As far as I know, previous versions of VBA-M, the original VBA and VBA-RR allow to configure up to 4 different sets of keys, leaving you the chance of changing between games or even setting a local multiplayer.

I have never tested it with a multiplayer, but I did some times set two different sets: one with my keyboard and the other one with my beloved Black European Gamecube controller (because we gotta remember that VBA and some of the forks of that time did not allow to assign more than one command to any button), so I could change it whenever I wanted it. Including this feature for this characteristic (and local multiplayer) would be pretty pointless, since you can assign several commands to a single button in the newer versions of VBA-M, but for using different sets for different games, sounds great.

Thanks for hearing me, and much welcomed for providing some help and ideas. If someone wants to tell Endrift about this, feel free to do.

P.S.: If someone wants a configuration for a Gamecube controller through a vJoy device, this is what I recommend (always preceded by "Joy1-", where Joynumber is your number of port, being 1 the one at your right):
A: Button0 (A)
B: Button1 (B)
L: Button6 (L)
R: Button5 (R)
START: Button7 (START)
SELECT: Button4 (Z)
Up: Button8 (D-Up) and/or Axis1- (A-Up)
Down: Button9 (D-Down) and/or Axis1+ (A-Down)
Left: Button10 (D-Left) and/or Axis0- (A-Left)
Right: Button11 (D-Right) and/or Axis0+ (A-Right)

**Comment by denisfa** [comment]
We could have a profile per game, that could be a `ini` file with the name of the game.
This could go like this, let's say, for game `loz.gba`.

1. Read `vbam.ini`;
2. Load game `loz.gba`;
3. Search and load `loz.ini` (somewhere for now, could be same dir, config dir etc)
4. When you quit/close the game, save all diff config to `loz.ini`.

This way it would be everything configurable per game, although it can also become a mess after some time.

As far as default options, I need to think more about it. I would say initially to just be empty and force users to configure directly on first run.

**Comment by thespbgamer** [comment]
I wanted, just like in VBAM V2.0.1, to be able to bind "KP_ADD" and "KP_SUBTRACT", now it's not even bindable, and allways changes the volume.

**Comment by denisfa** [comment]
@rkitover  see 36879ca8b49a102a76a81f663c23ffbfad2b53c3 for an idea of default keys.

@thespbgamer These keys will be available to bind on next release. The issue here is that they were hard-coded to volume changing and they were not available on the menu; so we could not do the rebinding. Now we can using `Options` > `Keys Shortcuts ...` > `Audio` > `{Increase, Decrease} Volume`. You unbind them from the default and then you can bind them to anything that you want.

**Comment by retro-wertz** [comment]
```
                    Suggested binds:
GBA keys    Keyboard         Gamepad(digital)    Extra(analog stick)
D-PAD         arrow-keys       d-pad                      bind to left-analog
A                 X                       button 0
B                 Z                       1
L                 A                        l1
R                S                        l2
SELECT     R-Shift                Select
START       Enter                   Start
```
```

**Comment by denisfa** [comment]
@retro-wertz What keyboard layout do you have in mind here? We were trying to support compact layouts. Also, any special reason for `L2` for `R`? I only fixed the joystick default keybinding (actually started to set it); I did not change it. `bind to left-analog` means two ways to use `D_PAD` with controller?

**Comment by rkitover** [comment]
@denisfa implemented WASD and KL/IO for game keys in #467 as well as 360 controller defaults. Works for me. Closing this now.

This will be in the next release.

**Comment by RunisDauphin** [comment]
For those of you that want to use the Retrobit Mega-Drive 6 button controllers to play here.
This are my two default configurations:
-/ L -> Z, R-> C:
Up=Joy5-Axis4-
Down=Joy5-Axis4+
Left=Joy5-Axis0-,Joy5-Axis1-,Joy5-Axis2-,Joy5-Axis3-
Right=Joy5-Axis0+,Joy5-Axis1+,Joy5-Axis2+,Joy5-Axis3+
A=Joy5-Button1
B=Joy5-Button2
L=Joy5-Button5
R=Joy5-Button4
Select=Joy5-Button8
Start=Joy5-Button9

-/ L -> X, R -> Y:
Up=Joy5-Axis4-
Down=Joy5-Axis4+
Left=Joy5-Axis0-,Joy5-Axis1-,Joy5-Axis2-,Joy5-Axis3-
Right=Joy5-Axis0+,Joy5-Axis1+,Joy5-Axis2+,Joy5-Axis3+
A=Joy5-Button1
B=Joy5-Button2
L=Joy5-Button0
R=Joy5-Button3
Select=Joy5-Button8
Start=Joy5-Button9

Hope you like it.

Runis Dauphin
